docs/source/migrate/index.rst

```diff
@@ -0,0 +1,274 @@
+
+  Usage: daml migrate TARGET_PATH SOURCE FROM_PATH TO_PATH
+
+  Available options:
+    TARGET_PATH              Path where the new project should be located
+    SOURCE                   Path to the main source file ('source' entry of the
+                             project configuration files of the input projects).
+    FROM_PATH                Path to the dar-package from which to migrate from
+    TO_PATH                  Path to the dar-package to which to migrate to
+    -h,--help                Show this help text
+
+For example, to create a migration project from ``foo-1.0.0`` to ``foo-2.0.0`` run
+
+.. code-block:: shell
+
+  daml migrate foo-upgrade-2.0.0 daml/Foo.daml foo-1.0.0/.daml/dist/foo-1.0.0.dar foo-2.0.0/.daml/dist/foo-2.0.0.dar
+
+This generates a migration project in the directory ``foo-upgrade-2.0.0``. To build it, change
+directory to ``foo-upgrade-2.0.0`` and run
+
+.. code-block:: shell
+
+  ./build.sh
+
+respectively ``.\build.cmd`` if your on a Windows system.  Note that you can **not** use the usual
+``daml build`` command to build the migration project.
+
+How migrations work and when it is necessary to write code manually
+-------------------------------------------------------------------
+
+It is important to understand that the ``daml migrate`` command will not always succeed in
+generating a migration project that will compile. To understand why, let's assume that the
+``foo-1.0.0`` consists of a single module:
+
+.. code-block:: shell
+
+  foo-1.0.0
+  ├── daml
+  │   └── Foo.daml
+  ├── daml.yaml
+  └── ui-backend.conf
+
+where the ``Foo.daml`` file contains
+
+.. code-block:: daml
+
+  daml 1.2
+  module Foo where
+
+  template Foo
+    with
+      a : Int
+      p : Party
+    where
+      signatory p
+
+The package ``foo-2.0.0`` contains exactly the same modules, but a new template ``Bar`` has been
+added to the ``Foo`` module.
+
+.. code-block:: daml
+
+  daml 1.2
+  module Foo where
+
+  template Foo
+    with
+      a : Int
+      p : Party
+    where
+      signatory p
+
+  template Bar
+    with
+      t : Text
+      p : Party
+    where
+      signatory p
+
+
+If we generate a migration project with ``daml migrate`` as above, the directory contents of the
+``foo-2.0.0-upgrade/daml`` directory is
+
+.. code-block:: shell
+
+  daml
+  ├── FooAInstances.daml
+  ├── FooBInstances.daml
+  └── Foo.daml
+
+For every template that was defined in the module ``Foo`` in the ``foo-1.0.0`` package, you will
+find two new templates. One to upgrade contract instances of this template to the changed template
+defined in the module ``Foo`` in the package ``foo-2.0.0`` and one to rollback the process. Here is
+an example:
+
+.. code-block:: daml
+
+   template FooUpgrade
+    with
+        op : Party
+    where
+        signatory op
+        nonconsuming choice Upgrade: ContractId B.Foo
+            with
+                inC : ContractId A.Foo
+                sigs : [Party]
+            controller sigs
```

Comment:
As soon as `sigs` contains more than one party calling this choice needs another template to collect the signatures. We should generate that one as well. We should also base as much of this as possible on generic templates in the near future.

Reply:
Agreed on the generic templates. I think we need some input anyways on how these upgrade templates should look like and what authorization they should have.

Reply:
yes, we should include a workflow to collect signatures, either via generic templates, or by generating an `UpgradeAgreement` template, which collects signatures and has non-consuming choices for `op` to create and  trigger `Upgrade` on any of the generated templates.